src/NServiceBus.Core.Tests/Pipeline/PipelineTest.cs

```diff
@@ -0,0 +1,358 @@
+﻿namespace NServiceBus.Core.Tests.Pipeline
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.IO;
+    using System.Linq;
+    using System.Linq.Expressions;
+    using System.Text;
+    using System.Threading.Tasks;
+    using ApprovalTests;
+    using Extensibility;
+    using NServiceBus.Pipeline;
+    using NUnit.Framework;
+    using Settings;
+    using Testing;
+    using FakeBuilder = Testing.FakeBuilder;
+
+    [TestFixture]
+    public class PipelineTest
+    {
+        [Test]
+        public async Task ShouldExecutePipeline()
+        {
+            var stringWriter = new StringWriter();
+
+            var pipelineModifications = new PipelineModifications();
+            pipelineModifications.Additions.Add(new Behavior1.Registration(stringWriter));
+            pipelineModifications.Additions.Add(new Stage1.Registration(stringWriter));
+            pipelineModifications.Additions.Add(new Behavior2.Registration(stringWriter));
+            pipelineModifications.Additions.Add(new StageFork.Registration(stringWriter));
+            pipelineModifications.Additions.Add(new Stage2.Registration(stringWriter));
+            pipelineModifications.Additions.Add(new Terminator.Registration(stringWriter));
+
+            var pipeline = new Pipeline<ITransportReceiveContext>(new FakeBuilder(), new SettingsHolder(), pipelineModifications);
+
+            var context = new TestableTransportReceiveContext();
+            context.Extensions.Set<IPipelineCache>(new FakePipelineCache());
+
+            await pipeline.Invoke(context);
+
+            Approvals.Verify(stringWriter.ToString());
```

Comment:
I'm not convinced that an approval test here is really adding value? Why not just making normal assertions? I don't think we expect constantly changing the values?